[Feat] 법정점검의 스케쥴을 추가, 상태 변경, 수정, 삭제 할 수 있다

Galmaeki opened the pull request:
# Pull Request 요약

- 관리자 화면의 일정 추가, 완료, 삭제, 수정 기능에 대한 api입니다

## 변경 사항

- Restdocs : 
  - 새 api 문서 작성
  - 유지관리 -> 법정점검으로 용어 변경
- ManagerScheduleController 추가
  - register, complete, delete, update 메소드 추가
- ScheduleConverter 클래스를 활용하도록 변경
- ManagerArticle : 일정 추가, 삭제 메소드 구현
- ManageSchedule : setter 추가
- ScheduleMustHaveOneException : 새 예외 추가
- ManageScheduleServiceTest : 테스트코드 추가
  - create 성공 테스트 1, 실패 테스트 3
  - complete 성공 테스트 1, 실패 테스트 3
  - update 성공 테스트 1, 실패 테스트 3
  - delete 성공 테스트 1, 실패 테스트 3

## 관련 이슈

- #133 

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [x] 테스트 코드 작성
- [x] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-05-31)
- 작업 종료일: (2024-06-01)

## 유의사항

- 일부 고민한 부분을 TODO로 남겨놨습니다! 머지하기 전 삭제 예정입니다! 조언 환영합니다!

Galmaeki committed:
[refactor] Schedule dto -> entity 컨버터 사용으로 변경함

변경 사항 :
- ManageSchedule : 컨버터에서 생성하도록 변경
- ManagerManageArticleService에서 컨버터 사용하도록 변경

Galmaeki committed:
[feat] 법정점검 Schedule 추가 코드 작성

변경 사항 :
- ManageArticle : 일정 단건 추가 기능 작성
- ManagerScheduleController, ManageScheduleService : 새 일정 생성 기능 추가

Galmaeki committed:
[feat] 예외 클래스 추가

변경 사항 :
- ScheduleNotFoundException : 예외코드, 예외 클래스 생성

Galmaeki committed:
[feat] 일정 완료처리 API 추가

변경 사항 :
- ManagerScheduleController, ManageScheduleService : complete 메소드 작성
- ManageSchedule : complete 메소드 작성
- ManageScheduleRepository : 같은 요청으로 다른 결과를 낼 수 있어 동시성에 대한 고려가 필요하여 우선적으로 비관적 락 사용, 추후 좋은 방식이 생길 경우 변경 예정....

Galmaeki committed:
[fix] 사용하지 않는 import 제거

변경 사항 :
- ManageSchedule : 사용하지 않는 import 제거

Galmaeki committed:
[feat] 일정 변경 api 작성

변경 사항 :
- ManagerScheduleController : update 메소드 작성
- ManageScheduleService :
  - update, updateSchedule 메소드 작성
- ManageSchedule : Setter 추가

Galmaeki committed:
[feat] 일정 삭제 api 작성

변경 사항 :
- ManagerScheduleController : delete 메소드 작성, 줄바꿈
- ManageScheduleService : delete 메소드 작성
- ScheduleMustHaveOneException : 예외 클래스 추가
- ErrorCode : SCHEDULE_MUST_HAVE_ONE 예외 코드 추가

Galmaeki committed:
[refactor] 삭제 로직 api 변경

문제 사항 :
- 엔티티 내부에 article이 존재하는걸 모르고 코드 작성 함

변경 사항 :
- ManagerScheduleController : 경로에서 받던 articleId 삭제
- ManageScheduleService : 검증로직 변경

Galmaeki committed:
[fix] 일정 생성 로직 변경

변경 사항 :
- ManageSchedule : 생성로직에서 연관관계 생성

Galmaeki committed:
[fix] 일정 삭제 로직 변경

변경 사항 :
- ManageArticle : 삭제시 연관관계 삭제하도록 수정
- ManageScheduleService : 삭제 로직 변경

Galmaeki committed:
[docs] 유지관리 명칭 변경 -> 법정점검

변경 사항 :
- manageArticle-api.adoc : 명칭 변경
- ManagerManageArticleControllerDocsTest : 명칭 변경

Galmaeki committed:
[docs] 일정 관련 API RestDocs 추가

변경 사항 :
- manageSchedule-api.adoc : 문서화
- ManagerScheduleControllerDocsTest : api 문서 생성
- index.adoc : 링크 추가

Galmaeki committed:
[fix] 성능을 위해 lock 제거

변경 사항 :
- repository : 성능을 위해 findAndPessimisticLockById 주석처리 - 추후 리팩토링 예정
- ManageScheduleService : 변경된 메소드 사용하지 않도록 수정

Galmaeki committed:
[fix] 잘못된 에러코드 변경

변경 사항 :
- ScheduleMustHaveOneException : 올바른 에러코드로 변경

Galmaeki committed:
[test] ManageScheduleServiceTest 추가

변경 사항 :
- ManageScheduleServiceTest :
  - create 성공 테스트 1, 실패 테스트 3
  - complete 성공 테스트 1, 실패 테스트 3
  - update 성공 테스트 1, 실패 테스트 3
  - delete 성공 테스트 1, 실패 테스트 3

Galmaeki committed:
[fix] 사용되지 않는 import 제거

변경 사항 :
- ManageScheduleRepository : 사용하지 않는 의존성 제거

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManageScheduleService.java:
```diff
@@ -0,0 +1,120 @@
+            .orElseThrow(ScheduleNotFoundException::new);
+
+        ManageArticle article = manageSchedule.getManageArticle();
+        //Todo article이 null일 수 있는지 확인
```
todo에 적어두신대로 article이 null인지 확인하는 예외 처리로직을 추가하면 좋을 것 같습니다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ManageScheduleService.java:
```diff
@@ -0,0 +1,120 @@
+        manageSchedule.complete();
+    }
+
+    public void update(
```
if문 대신 optional을 활용하면 코드가 훨씬 간결해질 것 같습니다!

Galmaeki replied:
어느 부분에 Optional을 사용해야할지 모르겠습니다!
조언해주실 수 있을까요?

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/controller/ManagerScheduleController.java:
```diff
@@ -0,0 +1,79 @@
+        return ResponseDTO.ok();
+    }
+
+    @PatchMapping("/progress/{manageScheduleId}")
```
스케줄의 일정변동과 함께 complete 상태만 따로 변경이 가능해야하니 PATCH를 유지하는게 더 좋을 것 같습니다!

ghrltjdtprbs reviewed (approved):
수고하셨습니다! 
멘토님도 말씀하셨다시피 메소드 이름은 간결한 것도 좋지만, 메소드명만 봐도 어떤 값을 받고 리턴할 것인지의 최소한의 정보는 포함되어야합니다(ex. scheduleCreate 등)
그리고 수정 로직에서 스케줄의 complete 상태 변경 로직이 누락되어있는 것 같은데 현재 pr에서는 제가 제대로 파악이 불가능하니 재현님께서 확인해주세요 ㅎㅎ 😊😊

asd42270 reviewed (approved):
아이고 고생 많으셨습니다!!

Galmaeki committed:
[style] 필요없는 주석 삭제

변경 사항 :
- ManagerScheduleController : 필요없는 주석 삭제

Galmaeki committed:
[fix] article 진행상황 변경 코드 추가

변경 사항 :
- ManageArticle :
  - checkSchedules 메소드 추가
  - 성능 개선을 위한 코드가 있으나 가독성이 심각하게 떨어져 보류함
- ManageScheduleService : 게시글 상태가 변경될 경우 위 메소드 실행하도록 변경

Galmaeki commented:
상태 변경로직 추가했습니다만 코드의 가독성과 성능에서 트레이드오프가 발생하여...의견 말씀해주시면 감사하겠습니다 😢

Galmaeki committed:
[refactor] checkSchedules 코드 성능 개선

변경 사항 :
- checkSchedules : 불필요한 반복문을 중간에 종료하여 성능을 개선
